lib/spack/spack/environment/environment.py

```diff
@@ -1345,28 +1345,38 @@ def concretize(self, force=False, tests=False):
             List of specs that have been concretized. Each entry is a tuple of
             the user spec and the corresponding concretized spec.
         """
-        if force:
-            # Clear previously concretized specs
-            self.concretized_user_specs = []
-            self.concretized_order = []
-            self.specs_by_hash = {}
-
-        # Remove concrete specs that no longer correlate to a user spec
-        for spec in set(self.concretized_user_specs) - set(self.user_specs):
-            self.deconcretize(spec)
-
-        # Pick the right concretization strategy
-        if self.unify == "when_possible":
-            return self._concretize_together_where_possible(tests=tests)
-
-        if self.unify is True:
-            return self._concretize_together(tests=tests)
-
-        if self.unify is False:
-            return self._concretize_separately(tests=tests)
+        old_concretized_user_specs = self.concretized_user_specs[:]
+        old_concretized_order = self.concretized_order
+        old_specs_by_hash = self.specs_by_hash
```

Comment:
Why is the array copied but the dicts not?

Reply:
Because I didn't sleep enough yesterday. I'll fix that, we should be copying all of them.

Reply:
Ok, figured out why the tests were passing even with this bug, and made the tests more robust in addition to fixing this.

In most cases, a concretization error would happen after the environment is cleared as part of the `concretize_*` methods or because `force=True`, in which case it doesn't matter whether we copy by reference or value. But if we copy by reference, there is a corner case if there is an internal Spack error raised before the environment is cleared, in which case we need to be copying by value.